pallets/token-mux/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+	pub type RatioFor<T> = <<T as Config>::Swaps as TokenSwaps<AccountIdFor<T>>>::Ratio;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config
+	where
+		CurrencyFor<Self>: From<LocalAssetId> + TryInto<LocalAssetId>,
+	{
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		type AssetRegistry: asset_registry::Inspect<
+			CustomMetadata = CustomMetadata,
+			AssetId = CurrencyFor<Self>,
+		>;
+
+		type Tokens: fungibles::Inspect<Self::AccountId> + Mutate<Self::AccountId>;
+
+		type Swaps: TokenSwaps<Self::AccountId, CurrencyId = CurrencyFor<Self>, Balance = BalanceFor<Self>>
+			+ OrderDetails<MuxSwap<CurrencyFor<Self>, RatioFor<Self>>, OrderId = OrderIdFor<Self>>;
+
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config>
+	where
+		CurrencyFor<T>: From<LocalAssetId> + TryInto<LocalAssetId>,
+	{
+		Deposited {
+			who: T::AccountId,
+			currency_out: CurrencyFor<T>,
+			currency_in: CurrencyFor<T>,
+			amount: BalanceFor<T>,
+		},
+		Burned {
+			who: T::AccountId,
+			currency_out: CurrencyFor<T>,
+			currency_in: CurrencyFor<T>,
+			amount: BalanceFor<T>,
+		},
+		SwapMatched {
+			id: OrderIdFor<T>,
+			amount: BalanceFor<T>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The given currency has no metadata set.
+		MetadataNotFound,
+		/// The given currency has no local representation and can hence not be
+		/// deposited to receive a local representation.
+		NoLocalRepresentation,
+		/// The given currency is not a local currency
+		NotLocalCurrency,
+		/// The provided local currency does not match the local representation
+		/// of the currency to be unlocked
+		LocalCurrencyMismatch,
+		/// Swap could not be found by id
+		SwapNotFound,
+		/// Matching orders does only work if there is a one-to-one conversion
+		NotIdenticalSwap,
+		/// This means the swap is either not a local to variant or not a
+		/// variant to local swap
+		InvalidSwapCurrencies,
+		/// Variant and local representation have mismatching decimals in their
+		/// metadata. A conversion between the two is not possible
+		DecimalMismatch,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		CurrencyFor<T>: From<LocalAssetId> + TryInto<LocalAssetId>,
+		BalanceFor<T>: FixedPointOperand,
+	{
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::deposit())]
+		pub fn deposit(
+			origin: OriginFor<T>,
+			currency_out: CurrencyFor<T>,
+			amount_out: BalanceFor<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let local = Self::try_local(&currency_out)?;
+
+			Self::mint_route(&who, local.clone(), currency_out.clone(), amount_out)?;
```

Comment:
Is `who` account coming from `liquidity-pools`? If so, I think nobody could call this.

Reply:
This call is permissionless, i.e. open to any signed origin, such that a bot can match those local swaps.

Reply:
The LP and FI flow is untouched by this pallet. It solely provides means of on- and off-ramping local stable coin variants by exposing the `match_swap(origin, order_id, amount)` extrinsic. This allows any signed extrinsic to fulfil any orderbook swap into/from `LocalAssetId`. The caller origin and the recipient of the swap are independent of each other.

Let's look at the current LP flow: An investor locks ETH USDC on Ethereum to invest into a DOT USDC pool (= `ForeignAsset(6)` on Centrifuge chain). When we receive the inbound message, we mint `LpEthUSDC` as foreign asset and create a swap for `LpEthUSDC -> ForeignAsset(6)`. Both currencies are publicly known and in order to fulfill the swap, someone needs to trade `ForeignAsset(6)` for `LpWrappedEthUsdc`.

The new flow will be adjusted slightly:
* USDC pools will have `CurrencyId::LocalAsset(LocalAssetId(1))` (:= `Local USDC`) as pool currency instead of DOT USDC
* The orderbook swap of `LpEthUSDC -> Local USDC` can be fulfilled by anyone who holds sufficient CFG to pay the transaction fee because this pallet transfers the `LpEthUSDC` amount to sovereign account of the token-mux pallet and mints the same amount as `Local USDC`
* For offramping (`Local USDC -> LpEthUSDC`), the corresponding amount needs to be in the sovereign account. Then, again anyone can fulfill the swap by transferring from sovereign to the swap order destination and burning the `Local USDC`